src/main/java/com/tabnine/inline/CompletionPreviewInsertionHint.kt

```diff
@@ -0,0 +1,91 @@
+package com.tabnine.inline
+
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.editor.LogicalPosition
+import com.intellij.openapi.editor.event.EditorMouseEvent
+import com.intellij.openapi.editor.event.EditorMouseEventArea
+import com.intellij.openapi.editor.event.EditorMouseMotionListener
+import com.intellij.util.Alarm
+import com.tabnine.inline.render.TabnineInlay
+import org.jetbrains.annotations.TestOnly
+import java.awt.Component
+import java.awt.Point
+import javax.swing.SwingUtilities
+
+private const val HINT_DELAY_MS = 100
+
+class CompletionPreviewInsertionHint(
+    private val editor: Editor,
+    private val inlay: TabnineInlay,
+    private var suffix: String
+) : Disposable,
+    EditorMouseMotionListener {
+    private var alarm: Alarm = Alarm(this)
+
+    init {
+        editor.addEditorMouseMotionListener(this)
+    }
+
+    override fun mouseMoved(e: EditorMouseEvent) {
+        alarm.cancelAllRequests()
+
+        if (inlay.isEmpty || e.area !== EditorMouseEventArea.EDITING_AREA) {
+            return
+        }
+
+        val mouseEvent = e.mouseEvent
+        val point = mouseEvent.point
+
+        if (isOverPreview(point)) {
+            alarm.addRequest(
+                {
+                    InlineHints.showPreInsertionHint(
+                        editor,
+                        SwingUtilities.convertPoint(
+                            mouseEvent.source as Component,
+                            point,
+                            editor.component.rootPane.layeredPane
+                        )
+                    )
+                },
+                HINT_DELAY_MS
+            )
+        }
+    }
```

Comment:
invert condition and early-return to avoid unnecessary indentation:

```suggestion
        if (!isOverPreview(point)) return

        alarm.addRequest(
            {
                InlineHints.showPreInsertionHint(
                    editor,
                    SwingUtilities.convertPoint(
                        mouseEvent.source as Component,
                        point,
                        editor.component.rootPane.layeredPane
                    )
                )
            },
            HINT_DELAY_MS
        )
    }
```